app/grandchallenge/algorithms/forms.py

```diff
@@ -622,3 +639,309 @@ class AlgorithmPublishForm(AlgorithmPublishValidation, ModelForm):
+            raise ValidationError("URL is not an algorithm detail view")
+
+        return {
+            "netloc": parsed_url.netloc,
+            "slug": resolver_match.kwargs["slug"],
+        }
+
+    def clean_algorithm_url(self):
+        algorithm_url = self.cleaned_data["algorithm_url"]
+
+        if Algorithm.objects.filter(
+            slug=self._parse_remote_algorithm_url(algorithm_url)["slug"]
+        ).exists():
+            raise ValidationError("An algorithm with that slug already exists")
+
+        return algorithm_url
+
+    def _build_algorithm(self, *, algorithm_slug, headers, netloc):
+        algorithms_list = self.remote_instance_client.list_algorithms(
+            slug=algorithm_slug, headers=headers, netloc=netloc
+        )
+
+        if algorithms_list["count"] != 1:
+            raise ValidationError(
+                f"Algorithm {algorithm_slug} not found, "
+                "check your URL and API token."
+            )
+
+        algorithm_serializer = AlgorithmSerializer(
+            data=algorithms_list["results"][0]
+        )
+
+        if not algorithm_serializer.is_valid():
+            raise ValidationError("Algorithm is invalid")
+
+        self.algorithm_serializer = algorithm_serializer
+
+    def _build_algorithm_image(self, headers, netloc):
+        algorithm_images_list = (
+            self.remote_instance_client.list_algorithm_images(
+                netloc=netloc,
+                headers=headers,
+                algorithm_pk=self.algorithm_serializer.initial_data["pk"],
+            )
+        )
+
+        algorithm_images = [
+            ai
+            for ai in algorithm_images_list["results"]
+            if ai["import_status"] == ImportStatusChoices.COMPLETED.label
+        ]
+        algorithm_images.sort(key=lambda ai: ai["created"], reverse=True)
+
+        if len(algorithm_images) == 0:
+            raise ValidationError(
+                "No valid algorithm images found for this algorithm, "
+                "check your URL and API token."
+            )
+
+        algorithm_image_serializer = AlgorithmImageSerializer(
+            data=algorithm_images[0]
+        )
+
+        if not algorithm_image_serializer.is_valid():
+            raise ValidationError("Algorithm image is invalid")
+
+        self.algorithm_image_serializer = algorithm_image_serializer
+
+    def _build_interfaces(self):
+        remote_interfaces = {
+            interface["slug"]: interface
+            for interface in chain(
+                self.algorithm_serializer.initial_data["inputs"],
+                self.algorithm_serializer.initial_data["outputs"],
+            )
+        }
+
+        self.new_interfaces = []
+        for slug, remote_interface in remote_interfaces.items():
+            try:
+                self._validate_existing_interface(
+                    slug=slug, remote_interface=remote_interface
+                )
+            except ObjectDoesNotExist:
```

Comment:
Don't you also need to catch the possible `ValidationError` that `_validate_existing_interface()` could throw?

Reply:
Ah no, you won't that to error out and not create a new interface in that case.